lib/src/config_provider/config_spec.dart

```diff
@@ -0,0 +1,811 @@
+import 'package:logging/logging.dart';
+import 'package:yaml/yaml.dart';
+
+final _logger = Logger('ffigen.config_provider.config');
+
+/// A container object for a ConfigSpec Object.
+class ConfigSpecNode<TE> {
+  /// The path to this node.
+  ///
+  /// E.g - ["path", "to", "arr", "[1]", "item"]
+  final List<String> path;
+
+  /// Get a string representation for path.
+  ///
+  /// E.g - "path -> to -> arr -> [1] -> item"
+  String get pathString => path.join(" -> ");
+
+  /// Contains the underlying node value after all transformations and
+  /// default values have been applied.
+  final TE value;
+
+  /// Contains the raw underlying node value. Would be null for fields populated
+  /// but default values
+  final Object? rawValue;
+
```

Comment:
Would it be useful to have `ConfigSpec` as a field here?

Maybe not, as we use it as an immutable data class with `copyWith`. All the logic is in other classes. Instead of a `ConfigValue` knowing how to transform itself because it has access to the config spec.

Reply:
Yeah, I intended this to act as an immutable data class. Maybe equivalent to `context` in some ways?